hphp/runtime/vm/jit/irgen-builtin.cpp

```diff
@@ -488,6 +473,113 @@ SSATmp* opt_set_frame_metadata(IRGS& env, uint32_t numArgs) {
+      case KindOfUninit:
+        return make_tv<KindOfUninit>();
+      case KindOfNull:
+        return make_tv<KindOfNull>();
+      case KindOfBoolean:
+        return make_tv<KindOfBoolean>(tmp->boolVal());
+      case KindOfInt64:
+        return make_tv<KindOfInt64>(tmp->intVal());
+      case KindOfDouble:
+        return make_tv<KindOfDouble>(tmp->dblVal());
+      case KindOfPersistentString:
+        return make_tv<KindOfPersistentString>(tmp->strVal());
+      case KindOfPersistentArray:
+        return make_tv<KindOfPersistentArray>(tmp->arrVal());
+      case KindOfPersistentVec:
+        return make_tv<KindOfPersistentVec>(tmp->vecVal());
+      case KindOfPersistentDict:
+        return make_tv<KindOfPersistentDict>(tmp->dictVal());
+      case KindOfPersistentKeyset:
+        return make_tv<KindOfPersistentKeyset>(tmp->keysetVal());
+
+      case KindOfVec:
+      case KindOfDict:
+      case KindOfKeyset:
+      case KindOfClass:
+      case KindOfString:
+      case KindOfArray:
+      case KindOfObject:
+      case KindOfResource:
+      case KindOfRef:
+        break;
+    }
+    // Other Kinds are not expected to be ConstVal
+    not_reached();
+  };
+
+  if (!func->isFoldable()) return nullptr;
+
+  // Don't pop the args yet---if the builtin throws at compile time (because
+  // it would raise a warning or something at runtime) we're going to leave
+  // the call alone.
+  PackedArrayInit args(numArgs);
+  for (auto bcOff = BCSPRelOffset{static_cast<int32_t>(numArgs - 1)};
+       bcOff > 
+         BCSPRelOffset{static_cast<int32_t>(numArgs - numNonDefaultArgs - 1)};
+       bcOff--) {
+    auto const t = topType(env, bcOff);
+    if (!t.hasConstVal() && !t.subtypeOfAny(TUninit, TInitNull, TNullptr)) {
+      return nullptr;
+    } else {
+      args.append(cellAsCVarRef(constAsCell(topC(env, bcOff))));
+    }
+  }
+  if (numArgs != func->numNonVariadicParams()) {
+    if (!topType(env).hasConstVal()) return nullptr;
+
+    auto const variadicArgs = constAsCell(topC(env)).m_data.parr;
+    auto const numVariadicArgs = variadicArgs->size();
+    for (auto i = 0; i < numVariadicArgs; i++) {
+      args.append(variadicArgs->get(i));
+    }
+  }
+
+  try {
+    // We don't know if notices would be enabled or not when this
+    // function would normally get called, so be safe and don't
+    // optimize any calls that COULD generate notices.
+    ThrowAllErrorsSetter taes;
+
+#ifdef DEBUG
+    VMRegAnchor _;
+    auto const savedPC = vmpc();
+    // ensure that vmpc is valid for later asserts within the invoke
+    vmpc() = vmfp()->m_func->getEntry();
+    SCOPE_EXIT{ vmpc() = savedPC; };
+#endif
+    RID().setJitFolding(true);
+    SCOPE_EXIT{ RID().setJitFolding(false); };
+
+    Cell retVal;
+    g_context->invokeFunc(&retVal, func, args.toArray(), nullptr, nullptr,
+                          nullptr, nullptr, ExecutionContext::InvokeNormal,
+                          !func->unit()->useStrictTypes());
+
```

Comment:
When we start jitting in separate threads (real soon now), this will almost certainly break. It might already break due to some of the guards that have recently been added to try to enforce that...